File: school/LAB/Lab8/process.pl
#!/usr/bin/env perl
use strict; use warnings; use v5.26;
use feature qw( signatures );
no warnings qw( experimental::signatures );

use Carp            qw( croak carp );
use Data::Dumper    qw( Dumper     );
use List::Util      qw( sum        );
use List::MoreUtils qw( minmax     );

###############################################################################
# Evil globals

our @stack;
our @level_stack;
our $TAB    = ' ' x 4;
our $output = '';
our $level  = 0;

###############################################################################
# File and data processing

sub process_file($filename) {
    my @grade_list;
    open('fh', '<', $filename) or croak $!;

    while (<fh>) {
        next if /^$/;
        chomp;
        my ( $student, $grd ) = split;
        push @grade_list, int($grd);
    }
    close('fh') or carp $!;

    return( sort( { $a <=> $b } @grade_list ) );
}

sub filter_data($grade_list) {
    my @array  = [];
    my $max    = 9;

    foreach my $grade (@{$grade_list}) {
        if ( $grade > $max ) {
            $array[ ++$#array ] = [];
            $max += 10;
            $max = 100 if $max == 99;
        }
        push @{$array[-1]}, $grade;
    }
    return @array;
}

sub print_data($average, $min, $max, $list) {
    print "Average: $average\nMax: $max\nMin: $min\n";
    $_ = join('', map({"[@{$_}]\n"} @{$list}));
    s/(\d+)(?![\]\d])/$1,/gm;
    print;
}

###############################################################################
# HTML generation functions

sub add_content($content) {
    my @stuff = split(/\n/, $content);
    foreach (@stuff) {
        chomp;
        $output .= $TAB x $level . "$_\n";
    }
}

sub check_stack {
    if ( $stack[-1] ) {
        add_content(pop @stack);
        push @stack, '';
    }
}

sub html($op) {
    if ( $op =~ /init/i ) {
        $output = "<html>\n\n";
    } elsif ( $op =~ /end|finish/i ) {
        $output .= "</html>\n";
    } else {
        carp 'Invalid operation.';
    }
}

sub do_add($sec) {
    push @stack, $sec;
    push @stack, '';
    push @level_stack, $level;
}

sub add(@ARGS) {
    check_stack;
    foreach my $sec (@ARGS) {
        do_add $sec;
        $output .= ${TAB} x $level++ . "<${sec}>\n";
    }
}

sub css($sec, $format) {
    check_stack;
    do_add $sec;
    $output .= ${TAB} x $level++ . "<${sec} ${format}>\n";
}

sub end($max = 1) {
    for ( 0 .. --$max ) {
        my $content = pop @stack;
        my $section = pop @stack;

        if ( $content ) {
            add_content $content;
        }
        $level   = pop @level_stack;
        $output .= ${TAB} x ${level} . "</${section}>\n";
        $output .= "\n" unless @level_stack;
    }
}

sub lit($in) {
    $stack[-1] .= $in;
}

sub para($in) {
    #$in =~ s|\n|</br>\n|g;
    $in =~ s/\n/<\/br>\n/g;
    $in = "$in\n" unless $in =~ /\n$/;
    $stack[-1] .= "<p>\n${in}</p>\n";
}


###############################################################################
# Main data processing


my @grade_list     = process_file('input.txt');
my $average        = sum(@grade_list) / scalar(@grade_list);
my ( $min, $max )  = minmax(@grade_list);
my @sectioned_data = filter_data(\@grade_list);

#print_data $average, $min, $max, \@sectioned_data;

# Perl lacks a way to generate a range with a step, so we do this the hard way.
my @ranges;
for ( my $I = 0; $I <= 90; $I += 10 ) {
    my $J = ($I == 90) ? (100) : ($I+9);
    push @ranges, sprintf('[%2d-%-3d]', $I, $J);
}
my $buf = join ' ', @ranges;


###############################################################################
# HTML page generation

html 'init';
add 'body';
    add 'h1';
        lit "This took so ridiculously long it's seriously embarrassing.";
    end;
end;

add 'body';
    para <<~ "EOF";
        Class average: $average
        Highest grade: $max
        Lowest grade : $min
        EOF
    add 'table';
        add 'tr';
            my $i = 13;
            foreach my $section (@sectioned_data) {
                my $len = scalar(@{$section}) * 35;
                css 'td', 'valign="bottom"';
                    lit <<~ "EOF";
                        <div style="width:62px;height: ${len}px ;background:blue; border:1px solid red; writing-mode:tb-rl"></div>
                        EOF
                end;
                css 'td', 'valign="bottom"';
                    lit <<~ "EOF";
                        <div style="width:${i}px;height: ${len}px ;background:white;"></div>
                        EOF
                end;
                ++$i;
            }
        end;
    end;
    css 'font', "face=courier size='3'";
        lit "$buf";
    end;
end;
html 'end';

###############################################################################
# And at long last:

print $output;

File: school/LAB/Lab8/Process.py
#!/usr/bin/env python3

import sys
FILE = 'input.txt'

##############################################################################
# Classes


class Data:
    def __init__(self, data):
        self.grades = data
        self.maxi = max(self.grades)
        self.mini = min(self.grades)
        self.mean = sum(self.grades) / len(self.grades)

        self.sections = self.make_sections()

    def make_sections(self):
        array = [[]]
        secmax = 9
        ind = 0
        for grade in self.grades:
            if grade > secmax:
                secmax = (secmax + 10) if (secmax != 99) else (100)
                array.append([])
                ind += 1
            array[ind].append(grade)

        return array

    def __str__(self):
        s = self
        buf = "Average: %s\nMax: %s\nMin: %s\n" % (s.mean, s.maxi, s.mini)
        for sec in self.sections:
            buf += str(sec) + "\n"
        return buf[:-1]


##############################################################################


class HTML:
    TAB = 4 * ' '

    def __init__(self):
        self.output = "<html>\n\n"
        self.stack = []
        self.level_stack = []
        self.level = 0

    def __add_content(self, content):
        stuff = content.split("\n")
        for line in stuff:
            line.rstrip()
            self.output += (self.TAB * self.level) + line + '\n'

    def __check_stack(self):
        if self.stack:
            self.__add_content(self.stack.pop())
            self.stack.append('')

    def __do_add(self, sec):
        self.stack.append(sec)
        self.stack.append('')
        self.level_stack.append(self.level)

    def add(self, lst):
        self.__check_stack()
        for sec in lst:
            self.__do_add(sec)
            self.output += (self.TAB * self.level) + "<%s>\n" % sec
            self.level += 1

    def css(self, sec, fmt):
        self.__check_stack()
        self.__do_add(sec)
        self.output += (self.TAB * self.level) + "<%s %s>\n" % (sec, fmt)
        self.level += 1

    def end(self, num=None):
        num = (1) if (num is None) else (num)
        for _ in range(num):
            content = self.stack.pop()
            section = self.stack.pop()

            if content:
                self.__add_content(content)
            self.level = self.level_stack.pop()

            self.output += (self.TAB * self.level) + "</%s>\n" % section
            if self.level_stack == 0:
                self.output += '\n'

    def lit(self, content):
        self.stack[-1] += content

    def para(self, content):
        content = "\n".join([i + "</br>" for i in content.splitlines()])

        if content[-1] != '\n':
            content += '\n'
        self.stack[-1] += "<p>\n%s</p>\n" % content

    def finish(self):
        self.output += "</html>\n"


##############################################################################
# Functions


def get_data():
    try:
        with open(FILE, 'r') as file_handle:
            data = [i.rstrip().split() for i in file_handle]
            # data = list(filter(lambda x: len(x), data))
            data = list(filter(len, data))
            return sorted([int(i[1]) for i in data])
    except FileNotFoundError as err:
        print(err)
        sys.exit()


##############################################################################
# Main

DATA = Data(get_data())
# print(str(DATA))

RANGES = ["[%2d-%-3d]" % (i, i+9) for i in range(0, 81, 10)]
RANGES.append("[90-100]")

H = HTML()

H.add(['body', 'h1'])
H.lit("This took so ridiculously long it's seriously embarrassing.")
H.end(2)

H.add(['body'])
H.para("Class average: %d\nHighest grade: %d\nLowest grade: %d\n"
       % (DATA.mean, DATA.maxi, DATA.mini))

H.add(['table', 'tr'])
width = 13

for sect in DATA.sections:
    length = len(sect) * 20
    H.css('td', 'valign="bottom"')
    H.lit("""\
<div style="width:62px;height: %dpx ;background:blue; border:1px solid red; writing-mode:tb-rl">
</div>""" % length)
    H.end()

    H.css('td', 'valign="bottom"')
    H.lit("""<div style="width:%dpx;height: %dpx ;background:white;">\n</div>""" % (width, length))
    H.end()
    width += 1

H.end(2)
H.css('font', "face=courier size='3'")
H.lit(" ".join(RANGES))
H.end(2)

H.finish()
print(H.output, end="")
